pinot-core/src/test/java/org/apache/pinot/core/data/manager/TableIndexingTest.java

```diff
@@ -0,0 +1,717 @@
+import org.apache.pinot.spi.data.DimensionFieldSpec;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.FieldSpec.DataType;
+import org.apache.pinot.spi.data.MetricFieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.data.readers.GenericRow;
+import org.apache.pinot.spi.utils.JsonUtils;
+import org.apache.pinot.spi.utils.builder.TableConfigBuilder;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+import org.apache.pinot.util.TestUtils;
+import org.jetbrains.annotations.NotNull;
+import org.testng.Assert;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+
+// Tests various combinations of field type, encoding, single/multi-value and index type
+// and compares test case results with TableIndexingTest.csv
+public class TableIndexingTest {
+  private static final File TEMP_DIR = new File(FileUtils.getTempDirectory(), "TableIndexingTest");
+  private static final String TABLE_NAME = "mytable";
+  private static final String OFFLINE_TABLE_NAME = TableNameBuilder.OFFLINE.tableNameWithType(TABLE_NAME);
+  public static final String COLUMN_NAME = "col";
+  public static final String COLUMN_DAY_NAME = "$col$DAY";
+  public static final String COLUMN_MONTH_NAME = "$col$MONTH";
+  public static final String COLUMN_WEEK_NAME = "$col$WEEK";
+
+  private final ArrayList<Schema> _schemas = new ArrayList<>();
+  private TestCase[] _testCases;
+  private Map<TestCase, TestCase> _testCaseMap;
+  private final List<TestCase> _allResults = new ArrayList<>();
+
+  @BeforeClass
+  public void setUp()
+      throws Exception {
+    TestUtils.ensureDirectoriesExistAndEmpty(TEMP_DIR);
+    createSchemas();
+    createTestCases();
+    readExpectedResults();
+  }
+
+  private void createTestCases() {
+    String[] indexTypes =
+        {
+            "timestamp_index", "bloom_filter", "fst_index", "h3_index", "inverted_index", "json_index",
+            "native_text_index", "text_index", "range_index", "startree_index", "vector_index"
+        };
+
+    _testCases = new TestCase[_schemas.size() * indexTypes.length];
+    _testCaseMap = new HashMap<>();
+
+    for (int i = 0; i < _schemas.size(); i++) {
+      for (int j = 0; j < indexTypes.length; j++) {
+        TestCase testCase = new TestCase(_schemas.get(i).getSchemaName(), i, indexTypes[j]);
+        _testCases[i * indexTypes.length + j] = testCase;
+        _testCaseMap.put(testCase, testCase);
+      }
+    }
+  }
+
+  private void readExpectedResults()
+      throws IOException {
+    List<String> expected = readExpectedFromFile();
+    // parse csv lines, e.g. INT;sv;raw;timestamp_index;true;
+    for (int i = 1; i < expected.size(); i++) {
+      String line = expected.get(i);
+      if (line.length() == 0) {
```

Comment:
(nit) My IDE usually gives warning on this
```suggestion
      if (line.isEmpty()) {
```